### DISCUSSION TITLE
How to avoid downtime using PM2

### DISCUSSION THREAD

**Discussion by yuriescl** [vercel/next.js#30899] [Help]
I'm deploying my Next.js app in a Ubuntu Server machine using PM2 (`pm2 start ...`), but every time I run `next build`, I get Internal Server Error, I imagine that's because the app cannot be running at the same time it's building.
I'd like some suggestions on how to build and restart the app without downtime. PM2 has cluster mode but it also fails when I run `next build` with the app running.

- Should I use `next export` to a directory, and build the app in another directory before replacing it?
- What happens if I build the app as a PWA, any suggestions on how to avoid downtime?

I'm aware that Vercel has built-in support for 0-second-downtime, just wondering how I'd do that in Ubuntu directly.

**Comment by ghost** [comment]
You can use Capistrano-like deploy with PM2 https://pm2.keymetrics.io/docs/tutorials/capistrano-like-deployments

**Comment by Kamleshpaul** [comment]
i solve this by create a script 

https://gist.github.com/Kamleshpaul/d22887c48e0c0d231734da24e8a80660

here to update my server i run this `sh nextjs-pm2-deploy.sh` it generate build on temp then replace with `.next` file and reload pm2

Here is article i wrote how it works
https://kamleshpaul.com/posts/next-js-deployment-script-for-zero-downtime-on-vps-with-pm2

**Reply by GregoryWullimann** [reply]
Great script!

I've noticed that sometimes it fails when moving `temp` to `.next`, the error is `rm: cannot remove '.next': Directory not empty`

I guess it happens because some file are created in `.next` between `rm -rf .next` and the `mv` command.

I think one solution is to create the `.next` folder if it doesn't exists, and then move everything from `temp` to `.next`, it will be a little bit slower since it's moving the files and not renaming the folder (_I think_).

```
rm -rf .next

if [ ! -d ".next" ]; then
  mkdir .next
fi
mv temp/* .next

rm -rf temp
```

**Reply by Kamleshpaul** [reply]
@GregoryWullimann  thank you.

can you tell me how can i create this error so i can test it.
i m using this in long time not getting any issue

**Reply by GregoryWullimann** [reply]
I wasn't able to reproduce it by running the commands individually, but this always happens in this way:

- Start script
- Go on website and keep refreshing a page (this will create files in .next) while the script is running
- Should get the error

**Reply by Kamleshpaul** [reply]
@GregoryWullimann i tried it all works

may be i have enable PWA in my website so if it broke it get from cache

**Reply by GregoryWullimann** [reply]
I tried it again on the live website, to be specific the error is
```
rm: cannot remove '.next/server/pages': Directory not empty
rm: cannot remove '.next/cache/images': Directory not empty
```

In this way the website will be down (error 502), and if I run again the script it works, since no file will be created in `.next` since the website is down.

Guess if anyone has a similar issue my fix would work. Otherwise yours should be a bit faster!

**Comment by marco910** [comment]
> i solve this by create a script 
> 
> 
> 
> https://gist.github.com/Kamleshpaul/d22887c48e0c0d231734da24e8a80660
> 
> 
> 
> here to update my server i run this `sh nextjs-pm2-deploy.sh` it generate build on temp then replace with `.next` file and reload pm2

@Kamleshpaul In which directory do you run the script. The root path of the Next.js project?

**Reply by Kamleshpaul** [reply]
yes i run this in root folder, i keep this `nextjs-pm2-deploy.sh ` file in root

**Comment by DEGA-hoangmai** [comment]
Hi @Kamleshpaul,

BUILD_DIR=temp is not working on my side. It doesn't create the temp dir; the output is still in .next. Do you have any special setup in package.json or any config files? Thank you.

**Reply by smith1302** [reply]
For anyone reading, make sure you set this line in your in your next.config.js:

`distDir: process.env.BUILD_DIR || '.next',` 

This is so NextJS can read from the `BUILD_DIR=temp` and build into that temp folder.

**Reply by isaacdarcilla** [reply]
Works for me.

**Comment by Nithur-M** [comment]
Hi,

It's 2025, is there a better solution for this?

**Reply by nihat-js** [reply]
Still no better and native solution
